src/rulesets/Base/array.jl

```diff
@@ -4,6 +4,10 @@
 
 ChainRules.@non_differentiable (::Type{T} where {T<:Array})(::UndefInitializer, args...)
 
+function frule((_, xdot), ::Type{T}, x::AbstractArray) where {T<:Array}
+    return T(x), T(xdot)
```

Comment:
Why not
```suggestion
    return T(x), xdot
```
?

By the same logic we use for `ProjectTo` which has that e.g. a `Diagonal` tangent is fine for a `Array` primal.

Am I correct in saying the problem with that would be that we actually need it to be dense.
Because we support mutation in forward mode so it must actually allocate memory to match?

Reply:
I have not thought super-hard about this, but my impression was they ought to be much more in lockstep than reverse mode. As you say, mutation is one reason.

Collections containing some `ZeroTangent`s showed up in trying to test Diffractor, I think from something like `x -> [x, 1]`. I guess the collection there is a tuple of arguments, not a primal collection.

There is a chance that demanding the same eltype here is too strict. What if you want to run ForwardDiff over CR forward mode? At least it'll be an error...